hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCancelAirdropHandler.java

```diff
@@ -0,0 +1,140 @@
+import com.hedera.node.app.spi.workflows.PreHandleContext;
+import com.hedera.node.app.spi.workflows.TransactionHandler;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.HashSet;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class contains all workflow-related functionality regarding {@link HederaFunctionality#TOKEN_CANCEL_AIRDROP}.
+ * This transaction type is used to cancel an airdrop which is in pending state.
+ */
+@Singleton
+public class TokenCancelAirdropHandler extends BaseTokenHandler implements TransactionHandler {
+
+    private final int MAX_ALLOWED_PENDING_AIRDROPS_TO_CANCEL = 10;
+
+    @Inject
+    public TokenCancelAirdropHandler() {
+        // Exists for injection
+    }
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenCancelAirdropOrThrow();
+        final var allPendingAirdrops = op.pendingAirdrops();
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+
+        for (final var airdrop : allPendingAirdrops) {
+            validateTrue(airdrop.hasSenderId(), INVALID_ACCOUNT_ID);
+            verifyAndRequireKeyForSender(airdrop.senderIdOrThrow(), context, accountStore);
+        }
+    }
+
+    /**
+     * Verifies that the sender account exists and ensures the account's key is valid and required for the transaction.
+     *
+     * @param senderId The AccountID of the specified sender whose key needs to be validated.
+     * @param context The PreHandleContext providing transaction context.
+     * @param accountStore The store to access readable account information.
+     * @throws PreCheckException If the sender's account is immutable or the sender's account ID is invalid.
+     */
+    private void verifyAndRequireKeyForSender(
+            @NonNull final AccountID senderId,
+            @NonNull final PreHandleContext context,
+            @NonNull final ReadableAccountStore accountStore)
+            throws PreCheckException {
+        final var senderAccount = accountStore.getAliasedAccountById(senderId);
+        validateTruePreCheck(senderAccount != null, INVALID_ACCOUNT_ID);
+
+        // If the sender account is immutable, then we throw an exception.
+        final var key = senderAccount.key();
+        context.requireKey(key);
+    }
+
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn, "Transaction body cannot be null");
+        final var op = txn.tokenCancelAirdropOrThrow();
+
+        validateFalsePreCheck(op.pendingAirdrops().isEmpty(), EMPTY_PENDING_AIRDROP_ID_LIST);
+        validateFalsePreCheck(
+                op.pendingAirdrops().size() > MAX_ALLOWED_PENDING_AIRDROPS_TO_CANCEL, MAX_PENDING_AIRDROP_ID_EXCEEDED);
```

Comment:
Max allowed pending airdrops should be a config

Reply:
If we want it in the config we should move the check in the handle, right? Since in the pure-check, we don't have access to the config. I can update the related handle story with this(https://github.com/hashgraph/hedera-services/pull/14380)

Reply:
yes, we should do it in handle and remove this check

Reply:
Added in the handle - https://github.com/hashgraph/hedera-services/pull/14380/commits/b0071b437b3e8261ebe739d9fa241468184300f2 👍